Grid block in core + primitive Container block (#3180)
plone/volto: docs/source/blocks/core/grid.md

---
myst:
  html_meta:
    "description": "Volto Grid Block description and developer notes"
    "property=og:description": "Volto Grid Block description and developer notes"
    "property=og:title": "Volto Grid Block"
    "keywords": "Volto, React, blocks, grid, container, Plone"
---

(grid-block)=

# Grid Block

```{versionadded} Volto 17.0.0-alpha.10
```

> Volto comes with a Grid block, building upon the primitive of a container block, it is a block that contains other blocks.

Reviewer: I would not mention the container block in the first sentence; that's an implementation detail and not relevant to users of the block. You could mention it further down.

Author: ok, but remember that this is dev docs, not user docs.

Reviewer: I know, I was just thinking there's a difference between a developer who wants to configure the existing block and a developer who wants to build one like it.

I would at least rephrase it slightly to:
Volto comes with a Grid block, which is a block that contains other blocks arranged in a horizontal row. It is implemented using the Container block primitive.

The Grid block uses a fixed horizontal layout with a configurable maximum items per block, defaulting to four elements.

The Grid block can potentially contain any block type. However, the used blocks might need some styling work to adapt to the horizontal and small widths in the container. For this reason, only these four are included by default: `slate` (text), `image`, `listing` and `teaser`.

## Settings

The Grid has the following settings:

Headline

> : The headline is a heading that appears above the title of the Grid Block.

Reviewer: ```suggestion
: The headline is a heading that appears above the columns of the Grid Block.
```

## Block configuration

The Grid block has these configuration settings:

templates

> : It's a function to provide a list of available templates.

Reviewer: What is a template?

Author: I pushed a more detailed description. 

Now I'm thinking that the TemplateChooser maybe should be part of the particular block, not of the container... I will think about it.

Reviewer: @sneridagh I don't have a strong opinion on that. The new description is helpful and helped me realize how the templates could be useful for other kinds of containers.

It has the following signature: (type) => (intl) => TemplateData[]
See the `src/components/manage/Blocks/Grid/template.jsx` for more info.
```ts
type TemplateData = {
    image: Icon,
    id: string,
    title: string,
    blocksData: BlocksData,
}

type BlocksData = {
  blocks: {
    [key: string]: object;
  };
  blocks_layout: {
    items: string[];
  };
};
```

maxLength
: The maximum elements length allowed inside the grid block.
By default is four elements.

allowedBlocks

> : The internal `id` of the allowed blocks in the grid block.

Reviewer: ```suggestion
: Array of block types allowed in the grid block.
```

## Data adapter

The Grid block has a data adapter function that allows you to both tap into the changes in the settings and issue changes in other settings fields.
This is valuable in the Teaser block because it saves an internal cache of the target element.
If you select the target, these values are updated.
When you update the target, by default these values remain, but you can issue another behavior.

The data adapter function is defined in the block's setting `dataAdapter`.
You can override it and add your own function, if required.
The following is the default adapter.
You should stick to this signature in your custom adapters.

```js
import { isEmpty } from 'lodash';

export const GridBlockDataAdapter = ({
  block,
  data,
  id,
  onChangeBlock,
  value,
}) => {
  let dataSaved = {
    ...data,
    [id]: value,
  };
  if (id === 'href' && !isEmpty(value) && !data.title && !data.description) {
    dataSaved = {
      ...dataSaved,
      title: value[0].Title,
      description: value[0].Description,
      head_title: value[0].head_title,
    };
  }
  onChangeBlock(block, dataSaved);
};
```

## Migration from `@kitconcept/volto-blocks-grid`

> From Volto 17.0.0-alpha.10 the Grid block, as it's in the `@kitconcept/volto-blocks-grid` add-on version 7.x.x, is included in Volto core.

Reviewer: ```suggestion
The Grid block was included in Volto core starting from Volto 17.0.0-alpha.10. It is based on the `@kitconcept/volto-blocks-grid` add-on version 7.x.x.
```

By default is four elements.

allowedBlocks
: Array of block types allowed in the grid block.
This can be extended to include any of the registered blocks

blocksConfig
: It allows you to customize the blocks config available for the inner blocks.
You have to pass the whole existing Volto `blocksConfig` of the main configuration, then modify it given your needs and pass it down.
You could add different variations, schemaEnhancers, etc or remove them as well but only for blocks inside the grid block.
```js
config.blocks.blocksConfig.gridBlock.gridAllowedBlocks: ['teaser', 'image', 'slate'];
config.blocks.blocksConfig.gridBlock.blocksConfig: {
  // One could customize the blocks inside the grid like this:
  ...config.blocks.blocksConfig,
  teaser: {
    ...config.blocks.blocksConfig.teaser,
    variations: [
      {
        id: 'default',
        isDefault: true,
        title: 'Default',
        template: DefaultBody,
      },
      {
        id: 'variation2',
        title: 'variation #2',
        template: DefaultBody2,
      },
    ],
  },
};
```

## Block Image Styling customization

By default, the images inside a Grid block are using a 16/9 aspect ratio.

> It is configurable by setting a LESS variable `@grid-images-aspect-ratio` or setting a CSS property somewhere in your project: `--grid-images-aspect-ratio`.

Reviewer: And what if I don't want a fixed aspect ratio? Set it to `unset`?

Author: Yes, we can document this. But by doing it, then the cards look awful (every image with its own height) if you don't set one. I'd say let's be opinionated on this one, document the override.

> There is also another setting for the `object-position` CSS definitin:  `@grid-images-object-position` and the corresponding `--grid-images-object-position`.

Reviewer: ```suggestion
There is also another variable for the `object-position` CSS property:  `@grid-images-object-position` and the corresponding `--grid-images-object-position`.
```

---
myst:
  html_meta:
    "description": "Volto grid block description and developer notes"
    "property=og:description": "Volto grid block description and developer notes"
    "property=og:title": "Volto grid block"
    "keywords": "Volto, React, blocks, grid, container, Plone"
---

(grid-block)=

# Grid block

```{versionadded} Volto 17.0.0-alpha.10
```

Volto comes with a grid block.
It builds upon the {doc}`container`.
It is a block that contains other blocks.
The grid block uses a fixed horizontal layout with a configurable maximum number of blocks, with a default of four blocks.

The grid block can contain any block type.
However, due to its horizontal alignment and narrow widths of the blocks it contains, the blocks might need additional styling.
For this reason, only four blocks are included by default, namely `slate` (text), `image`, `listing`, and `teaser`.

## Settings

The grid block has the following settings.

`Headline`

> :   The headline is a heading that appears above the title of the grid block.

Reviewer: @stevepiercy It's not "above the title". It is the title. We already fixed this before your latest edits. (Thanks for those, btw.)
```suggestion
:   The headline is a heading that appears above the columns of the grid block.
```

It is configurable by setting a LESS variable, `@grid-images-aspect-ratio`, or setting a CSS property somewhere in your project, `--grid-images-aspect-ratio`.
There is another setting for the `object-position` CSS definition, `@grid-images-object-position`, and its corresponding `--grid-images-object-position`.

> If you want to disable this feature from the grid images, you should set both to `unset`.

Reviewer 2: Does "both" refer to the aspect ratio and the position? If so, then:

```suggestion
If you want to disable a default setting, then set its value to `unset`.
```
